### DISCUSSION TITLE
Is V-AC correct?

### DISCUSSION THREAD

**Discussion by wrenix** [Sn0w3y/NEP-Inverter-Reversing#15] [General]
My fetched binary code was:

`792600401400000f0f0f0f00001c00c3c3c3c3xxxxxxxx00002107cf14a00f0104a6315f0bca00058cbb936993`

as table:
```python
binary_data = bytes([
  #------#------#------#------#------#------#------#------#
    0x79,  0x26,  0x00,  0x40,  0x14,  0x00,  0x00,  0x0f, # 8
  #------#------#------#------#------#------#------#------#
    0x0f,  0x0f,  0x0f,  0x00,  0x00,  0x1c,  0x00,  0xc3, # 8
  #------#------#------#-------SERIAL-NUMBER-------#------#
    0xc3,  0xc3,  0xc3, sn[0], sn[1], sn[2], sn[3],  0x00, # 8
  #------#-V-AC-#-P-AC-#------#------#------#------#------#
    0x00,  0x21,  0x07,  0xcf,  0x14,  0xa0,  0x0f,  0x01, # 8
  #------#------#------#------#------#------#------#------#
    0x04,  0xa6,  0x31,  0x5f,  0x0b,  0xca,  0x00,  0x05, # 8
  #------#------#------#------#------#
    0x8c,  0xbb,  0x93,  0x69,  0x93 # 5
])
```
So:

V-AC:  is 0x00,0x21: 33V ?
P-AC:  is 0x07 *3.190: 22.33 W ?

**Comment by Sn0w3y** [comment]
We still need to verify some of the Values.. also we need to Reverse Engineer the Values.

Also there should AC, DC, AMPERE, WATT in the Payload i guess (also Status could be included)

But we still need to find out what is what - it was just assumptions :)

I will fire up my NEP later the day and try to reverse it :)

BTW: I like the way you work on this project and appreciate your work!

**Comment by wrenix** [comment]
Thank you for debugging all the data and merge my work -> i really (i like to pool forces).

I do not register (yet) on nepviewer - so i hope that my fetched data (like in this issue) helps to debug all the values correct.

Thats the reason why i spend this data (and write an universal Datapoint interpreter the models.py in #2).

**Comment by Sn0w3y** [comment]
@wrenix 
I will try to inject a fixed Wattage into the Inverter with a "DC-Labornetzteil" and with this i can see the EXACT Wattage aswell as the DC Voltage so i can see the exact values :)

Greetings

**Comment by wrenix** [comment]
The Data of today -> i strip the prefix (with serial_number away) and find a small pattern in the data:
```
cat /var/log/nepserver-payload.log | sed 's/792600401400000f0f0f0f00001c00c3c3c3c3.\{8\}0000\(............0.....31..0...0.0.8c........\)/\1/'
23020e14a00f0002a431700b5500058c958b0ed6
b4009611a00f0001ae31600b5800058cab862ff9
08000411a00f0001ac31500b5800058c5b858e16
00003f12a00f0000aa314f0b5800048c8a85ecee
00005611d00f00019e31d00a0000058cef84b462
01000e11a00f0001a431df0a0000058c4d7b5dd5
00004f11e00f0001a031ef0a0000058c198509bd
00003211c00f0001a431f00a0000058ca17b054b
00004f11d00f0001a831000b0000058c22851c50
02003f11900f0001a8310f0b0000058ca37b0ae4
e9002f1100100001a231100b0100058c2686260a
c500b211d00f0001a231200b0100058c257d104a
5201bd1110100002a231200b0300058c3f885c14
2b01c511e00f0001a2312f0b0300058cbf7d4575
6202df1210100002a231300b0800058c868aeee6
29028b13e00f0002a431400b0700058c9a81038f
1b029f1310100002b031400b0d00058c228b281c
b801f511e00f0001ac31400b0b00058c9f7e069c
bb021a1210100002b031400b1200058c428a6646
6f021712f00f0002b0314f0b0f00058c9f800b53
2a04751200100002b831500b1b00058c1d8c20e6
e203fb12e00f0002b6315f0b1700058c0483d9b5
2906491220100003ae31600b2800058c238f32d2
cc051113f00f0003aa316f0b2300058c2486205a
33079e14f00f0104a031700b3800058c6593bd85
04061a13c00f00049e317f0b3100058c478769f9
0c089a14f00f0105b231800b4900058c6d95d12d
0907ae13c00f0004b031800b4000058c0289e2aa
dc076f1400100105ac31900b5c00058c3f9575a1
69064614e00f0004b231900b5100058c3d895838
7c239a1500100211ae319f0b7800058c38beb9c3
4b21be15e00f0210b2319f0b6d00058cc4b1b620
6a24dc15e00f0212b631df0bc800058ce0c00c26
b41fdc14c00f020fb431cf0bb100058c31b4955d
4213be13b00f0109b631d00bb600058c089b1121
421c3415d00f010eb8311f0c2301058c9cb46eda
721f7a14b00f010fb831000c0401058c04b43d0b
101d0114b00f010eb631100c0e01058cb1aa8434
d511ca14d00f010994313f0c6f01058c0da53078
641e8c15c00f020f9631300c4a01058c6db40fa1
28316715c00f02189431300c5501058c85ca6b2f
b918e914e00f010cae31600cce01058cbaaf9e72
0a17a014c00f010cb031500ca101058cb7a98d81
41178b14c00f010bae315f0cad01058c62a3d6be
0210471500100108aa31600c0202058cb1a37759
6e0f6614e00f0108a831600cd101058c399e7afe
7a0e7e14e00f0106aa31600cdc01058cda95aa80
0513881510100109ac315f0c2702058c2ba8740e
3213ac14f00f010aaa31500cf501058c69a4e61a
a414d214d00f010bae31500c0002058c60a1ceb0
3b170a14f00f010bb231600c5d02058c00ad273f
4d15bf13d00f010ab4316f0c2a02058cd5a5c117
2b148b15e00f010ab231600c3102058cf5a0f7df
a616d414e00f010caa316f0c9202058c61aeead0
64158c13d00f010aac316f0c5a02058ce1a5d951
06152715c00f010baa31600c6202058c75a2fbb7
af123d15001001099a31700ccd02058c1fa85bc7
12131714e00f010a9c31700c9202058c53a4bbd9
97111515c00f01089e31600c9a02058c209c44b4
e811d41410100109a431700cfc02058c38a78a5c
0111e112f00f0108a431700cbd02058c4f9fa81a
e50e3715d00f0107a6316f0cc402058ce198be48
f3135f15b00f0109a631700c2603058c9aa95305
be13ae13900f0109a6317f0ce302058cb0a37218
e5145e14b00f010aa231700ceb02058c0ca125b7
f516af14e00f010aa831800c6103058c69adf8ea
45170e14c00f010ca631800c1a03058cffaa2054
ee145115b00f010aa631800c2203058c53a2b6d2
b81fd0140010010fa431900c9a03058cb2bba701
241efc13e00f020ea431900c5103058c32b49862
f42c4816e00f0216a831900c5903058ccac7ee3c
88087f15c00f0104aa31800cbb03058c089a10f8
4908a414a00f0104ac318f0c7003058ce494bdcf
2e081d15b00f0104ac317f0c7703058cb98f5d2b
a411f513f00f0109ac31600cdf03058cd0a6b81a
550fff13d00f0109aa31700c9103058c7ba00246
400f3015b00f0109a631600c9803058cda9bb7ef
5e1c5415e00f010ea431700c0c04058c0eb85951
73199514c00f010ca431700cbb03058c43aeae1c
a21b0d15b00f020ea4316f0cc203058c51adc8ca
8d157714e00f010aa8319f0c5004058cd7acd3df
1315af13c00f010aaa31900cfc03058c64a7e2cc
bd16a314a00f010caa31900c0204058c58a6c852
157b9a142010043caa31c00cc504058c5d451183
4462901500100430aa31c00c6a04058c571db57d
1f634b15d00f0432ac31cf0c7404058c481a90a4
8f2da81410100217a6314f0d9105058c2ad4c951
78503115f00f0328a6314f0d2805058c64049dc9
391f8d14f00f020fa8315f0d2e05058c08b44402
70183115d00f020ca4315f0d1306058ce4b4fefa
f616de13b00f020aa2315f0da105058cd3abc80e
54170415b00f020ca831600da605058cd3a9c5dd
2d1a0f15d00f010ca631100d4306058c61b6fc02
be190714c00f010ca631200dcf05058cc6b0b9f9
d5197115a00f020da8310f0dd305058c8dad4076
3423dd15e00f0211a431e00c8706058c59c508a2
6121fc14d00f0211a431ff0c0f06058c8cbe60b0
21204914c00f020fa831e00c1406058ce3b4fbbd
```

**Comment by Sn0w3y** [comment]
will try to deobfuscate it later on :)

**Comment by Sn0w3y** [comment]
I opened a Discussions Page for every Wattage Class i know so far - as it seems every Power Class has the Wattage in a different Spot (or it changes the location?)

Could you please (instead of removing the serial number) reset the serial to XX XX XX XX ?

https://github.com/Sn0w3y/NEP-Inverter-Reversing/discussions

**Comment by Sn0w3y** [comment]
@wrenix see https://github.com/Sn0w3y/NEP-Inverter-Reversing/blob/feature/scraping-ai/log.txt

**Comment by Sn0w3y** [comment]
@wrenix i found out, that the Data is encrypted. We need an Approach to Dump the Firmware off the LPT270 to see which Keys they use.

**Comment by javadwarf** [comment]
Hi, maybe the following document will help to further analyze the data: HF-LPX70 Series Wi-Fi_BLE Module User Manual-V2.7(20240126) from http://www.hi-flying.com/hf-lpt270 under Downloads. Regards Stefan

**Comment by DE-cr** [comment]
> @wrenix i found out, that the Data is encrypted. We need an Approach to Dump the Firmware off the LPT270 to see which Keys they use.

Looking at the data above (https://github.com/Sn0w3y/NEP-Inverter-Reversing/issues/4#issuecomment-2017785486), it doesn't look encrypted to me. There's too many patterns in it, e.g.
* byte 9 stays fixed at 31 (which could stand for 49 Hz, but I'd expect a float here, not an int)
* bytes 15+14 stay almost constant at 8c05, which reminds me of a field in my (rebranded BDM600) inverters' data returned from the NEP servers, which mostly reads 8a05 (but sometimes is 8a01 (usually early mornings), or less often 8a02/8a03/8a04).
* bytes 13+12 (almost) represent an increasing value, starting at zero and reaching 1671. This could be the day's total Wh for that system. (However, that sequence is not quite in order, with some rows shuffled around.)

When asking the NEP servers for data from a micro inverter for a particular day, one gets a series of samples containing twelve fields each: date, time, kW, kWh, V-DC, V-AC, Hz, °C, eng offset (?), ratio (?), ver (?), status (0000=okay). That data must have been sent from the inverter to the cloud, i.e. it will also be in your payload packages captured. (Well, date and time _could_ also not be sent, but assigned when the server receives the data.)

For me, I've decided to use the NEP cloud, and process the data my inverter sends there _locally_ (see https://github.com/DE-cr/NEPviewerCR) instead of using NEPviewer, but I still find this repository here interesting.

**Comment by Sn0w3y** [comment]
Hi @DE-cr,

Thanks for your insights! I agree, the data doesn't appear to be encrypted given the clear patterns visible in the payload. Here are some thoughts on your observations:

1. **Byte 9 (`31`):**
   - It's consistent across samples, and your assumption of it representing `49 Hz` aligns with common grid frequencies. While a float might typically represent frequency, this could be a fixed integer value for simplicity or internal scaling.

2. **Bytes 15+14 (`8c05`):**
   - The consistency you noted with `8a05` in your BDM600 data is intriguing, especially with variations like `8a01` in the early morning. This could indeed reflect a status or environmental condition such as inverter state or temperature thresholds.

3. **Bytes 13+12 (incrementing values):**
   - The interpretation as cumulative energy (`Wh`) makes sense, especially since they increase monotonically throughout the day. The slight shuffling of rows could point to asynchronous data collection or transmission order differences.

It’s interesting that you've decided to use the NEP cloud and process the data locally—it’s a practical approach, especially with tools like [NEPviewerCR](https://github.com/DE-cr/NEPviewerCR). While reverse-engineering the raw data here offers deeper insights into the protocol, your method ensures real-time usability for your system.

Looking forward to any further patterns or discoveries you identify, and thanks for contributing this perspective to the discussion! Maybe you could make PR's to this Project aswell? :)

Greetings!

**Comment by DE-cr** [comment]
What would make this repo's goal here much easier is having your system send to your local server, and from there forward the data to the NEP server as well, then fetch the processed data from the latter via https://user.nepviewer.com/pv_monitor/proxy/history/{sn}/{yyyymmdd}/{yyyymmdd}/0/1/2 and compare it to the raw payload intercepted earlier. Once you have that data, I'd gladly take another look. :)

**Comment by DE-cr** [comment]
Fyi: NEP has changed the format of their 'history' reports: The '8a05' ('ver'?) field is gone, and two more floats are now in each data set. Seems there's more interesting data in what the inverters send to the cloud. :)

**Comment by DE-cr** [comment]
Fyi: When NEP took down the API I had been using in my https://github.com/DE-cr/NEPviewerCR tools until April 2025, I decided to just use their official NEPViewer web site and app, as they have been improved a lot since when I bought my solar system. However, I've now revived NEPviewerCR - this time as a supplement to NEP's viewers, not a replacement.

**Comment by Sn0w3y** [comment]
> Fyi: When NEP took down the API I had been using in my https://github.com/DE-cr/NEPviewerCR tools until April 2025, I decided to just use their official NEPViewer web site and app, as they have been improved a lot since when I bought my solar system. However, I've now revived NEPviewerCR - this time as a supplement to NEP's viewers, not a replacement.

Very very nice work - just had a quick look at it - my Goal was to have it only locally as there are many Users out there which only wanted the Monitoring locally - i guess i will stop developing it as:

1. i still have not found out all values
2. other projects require more Attention right now :(
3. I am not using a NEP anymore so i also can not successfully Review changes except i have trustworthy PRs 

Either way - thanks for your work!!!

**Comment by ornkrist** [comment]
@wrenix how many watts are those panels  ?

**Comment by ornatwork** [comment]
Maybe something like this 

| Byte Position | Field Name | Format | Scaling | Value Range | Interpretation | Confidence |
|---------------|------------|--------|---------|-------------|----------------|------------|
| 0-18 | Unknown1 | 19 bytes | - | Variable | Unknown field - purpose unclear | Low |
| 19-22 | Serial Number | 4 bytes | - | 00 70 07 d1 | **CONFIRMED** - Serial number | High |
| 23 | Marker1 | Single byte | - | 0 | Marker byte | High |
| 24-27 | Unknown2 | 4 bytes | - | Bytes 0-1: 0-8225, Bytes 2-3: 4356-5516 | Unknown field - purpose unclear | Low |
| 28-29 | AC Voltage | Little-endian 16-bit | Divide by 16 | 249.0-258.0V | **CONFIRMED** - Matches gateway 250.3V | **Very High** |
| 30-31 | AC Power | Little-endian 16-bit | 0.1W units | 0.0-384.2W | **CONFIRMED** AC output power (after removing high test data > 400W) | High |
| 32 | DC Voltage | Single byte | Divide by 5 | 29.6-36.8V | **CONFIRMED** - Matches gateway 30.8V and 400W panel spec | **Very High** |
| 33 | Marker2 | Single byte | - | Always 0x31 | Fixed marker byte | **Very High** |
| 34-35 | DC Current | Little-endian 16-bit | Divide by 250 | 11.07-13.70A | **CONFIRMED** - Matches 400W panel spec (9-13A) | **High** |
| 36-37 | Temperature | Little-endian 16-bit | - | 0-1603 | **CONFIRMED** - Temperature | **Medium** |
| 38 | Status | Single byte | - | 4-5 | **CONFIRMED** - 0x04=Error, 0x05=Normal | **High** |
| 39 | Marker3 | Single byte | - | Always 0x8c | Marker byte | **Very High** |
| 40-44 | Unknown4 | 5 bytes | - | Variable | Unknown field (5 bytes) | Low |

**Comment by wrenix** [comment]
From @ornkrist 
> [@wrenix](https://github.com/wrenix) how many watts are those panels ?

My one, i do not know. Sorry this panels and inverters are in the roof of my parents (i just manage the rpi and the network there ... so sadly i am not able to reverse-enegier everything).

PS: Thank you in alle the nice work of you @DE-cr and @ornatwork .

**Comment by Sn0w3y** [comment]
Also from me MANY thanks for the effort ! I would love to see PRs to this Project as i do not use these Inverters anymore - but maybe we could bring a little more to the Table with your help ! :)

**Reply by ornatwork** [reply]
I could dig around some more if I can get a full 45 byte payload dump, more data the better.

**Reply by Sn0w3y** [reply]
Sure, unfortunately I can not provide this but maybe @wrenix could?

**Reply by wrenix** [reply]
With the new software? I have to look how to update.

**Reply by ornatwork** [reply]
The codebase has not been updated, the software is the same as before.

If you can capture full payloads (  45 bytes  ) - including serial numbers ( xxxxxxxx ) that would help a lot, then I can research it further.

792600401400000f0f0f0f00001c00c3c3c3c3xxxxxxxx00002107cf14a00f0104a6315f0bca00058cbb936993
.
.
.
